#!/usr/bin/Rscript

#### Title: Numbat script for melanoma
#### Author: Jana Biermann, PhD

library(numbat)
library(dplyr)
library(Seurat)
library(ggplot2)
library(glue)
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(ggtree))
library(stringr)
library(tidygraph)
library(patchwork)
'%notin%' <- Negate('%in%')

pat<- commandArgs()[6]

mypal = c('1' = 'gray', '2' = "#377EB8", '3' = "#4DAF4A", '4' = "#984EA3", 
          '5'="#ff9768", '6'='#ae1717', '7'='#0f04b5', '8'="#f87382",
          '9'='green','10'='yellow','11'='deeppink','12'='black','13'='purple')


## Get count_mat for pat
print(paste('Processing:', pat))
system(paste0("aws s3 sync s3://melanoma-ribas/inferCNV/inferCNV_subcluster_",pat,"/ data/",pat,"/ --exclude '*' --include '*_cnv.rds' --quiet"))
seu<- readRDS(paste0("data/",pat,"/",pat,"_cnv.rds"))
count_mat<-seu@assays$RNA@counts
colnames(count_mat)<-seu$barcode_orig

## Get ref_mat from entire cohort
if(!file.exists('data/tumor/numbat/data_numbat_melanoma_internal_ref.rds')){
  system('aws s3 cp s3://melanoma-ribas/numbat/data_numbat_melanoma_internal_ref.rds data/tumor/numbat/ --quiet')
}

if(!file.exists('data/tumor/numbat/data_numbat_melanoma_internal_ref.rds')){
  if(!file.exists('data/merged/data_melanoma_merged_v2.rds')){
    system('aws s3 cp s3://melanoma-ribas/Seurat/merged/data_melanoma_merged_v2.rds data/merged/')
  }
  
  seu_all<-readRDS('data/merged/data_melanoma_merged_v2.rds')
  seu_all$annot<-ifelse(is.na(seu_all$celltype_bped_main)==T, 'unknown',seu_all$celltype_bped_main)
  notum<-subset(seu_all,annot %notin% c('Melanocytes','unknown'))
  ref_mat<-notum@assays$RNA@counts
  colnames(ref_mat)<-notum$barcode_orig
  cell_annot=data.frame(cell=notum$barcode_orig,group=notum$annot)
  ref_internal = aggregate_counts(ref_mat, cell_annot)
  ifelse(!dir.exists(file.path('data/tumor/numbat/')),
         dir.create(file.path('data/tumor/numbat/'),recursive = T), FALSE)
  saveRDS(ref_internal,'data/tumor/numbat/data_numbat_melanoma_internal_ref.rds')
}

ref_internal<-readRDS('data/tumor/numbat/data_numbat_melanoma_internal_ref.rds')

## allele dataframe generated by pileup_and_phase script
system(paste0('aws s3 cp s3://melanoma-ribas/numbat/',pat,'/',pat,'_allele_counts.tsv.gz data/',pat,'/ --quiet'))
df_allele<-read.delim(paste0('data/',pat,'/',pat,'_allele_counts.tsv.gz'))
df_allele<- subset(df_allele,cell %in% seu$barcode_orig)

## Run numbat
out = run_numbat(count_mat=count_mat, # gene x cell integer UMI count matrix 
                 lambdas_ref=ref_internal, # reference expression profile, a gene x cell type normalized expression level matrix
                 df_allele=df_allele, # allele dataframe generated by pileup_and_phase script
                 out_dir= paste0('data/',pat),
                 genome = "hg38",
                 ncores = 16,
                 ncores_nni=16,
                 plot = TRUE)


## Read numbat
nb = Numbat$new(out_dir = paste0('data/',pat))

# Single-cell CNV calls
cnv_calls<- nb$joint_post %>% select(cell, CHROM, seg, cnv_state, p_cnv, p_cnv_x, p_cnv_y)
table(cnv_calls$cnv_state,useNA = 'always')
cnv_calls %>% group_by(cnv_state) %>% arrange(p_cnv,desc=F)

# clone info
clones<-dim(table(nb$clone_post$clone_opt))
clone_info<-nb$clone_post
seu$cell<-seu$barcode_orig
seu$barcode_tmp<-rownames(seu@meta.data)
seu@meta.data<-left_join(seu@meta.data,clone_info,by='cell')
rownames(seu@meta.data)<-seu$barcode_tmp

## Get clone segments
segs<- nb$segs_consensus
segs<- segs %>% group_by(seg_cons) %>% select(seg_cons, sample, CHROM,seg_start, seg_end, cnv_states)
segs$sample<-ifelse(is.na(segs$sample)==T,NA,paste0(pat,'_clone',segs$sample))
segs$pat_seg<-paste0(pat,'_seg',segs$seg_cons)
bulk<-nb$bulk_clones
bulk<-bulk %>% 
  select(seg_cons,logFC) %>% 
  na.omit() %>%
  group_by(seg_cons) %>%
  summarise(log2FC_median = median(logFC),
            log2FC_mean = mean(logFC))
segs<-left_join(segs,bulk,by='seg_cons')
write.csv(segs,paste0("data/",pat,"/table_",pat,"_segs.csv"),row.names = F)

clon<- nb$clone_post %>% select(cell, clone_opt, GT_opt)
write.csv(clon,paste0("data/",pat,"/table_",pat,"_clone.csv"),row.names = F)


## plots
pdf(paste0("data/",pat,"/plots_",pat,"_numbat.pdf"),width = 10)
# Copy number landscape and single-cell phylogeny
nb$plot_phylo_heatmap(clone_bar = TRUE, p_min = 0.9,raster = T,pal_clone = mypal)

# Consensus copy number segments
nb$plot_consensus()

# Bulk CNV profiles
nb$bulk_clones %>% 
  filter(n_cells > 50) %>%
  plot_bulks(min_LLR = 10, # filtering CNVs by evidence
             legend = TRUE,raster=T)

# clones
DimPlot(seu, group.by = 'clone_opt',shuffle = T, raster=T,cols = mypal[1:clones])
DimPlot(seu, group.by = 'GT_opt',shuffle = T, raster=T)

# Tumor versus normal probability
p1<-FeaturePlot(seu, features  = c('p_cnv'),order = T, raster=T)+
  scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, name = 'Posterior')+
  ggtitle('Tumor vs normal probability\n(joint)')

p2<-FeaturePlot(seu, features  = c('p_cnv_x'),order = T, raster=T)+
  scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, name = 'Posterior')+
  ggtitle('Tumor vs normal probability\n(gex)')

p3<-FeaturePlot(seu, features  = c('p_cnv_y'),order = T, raster=T)+
  scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, name = 'Posterior')+
  ggtitle('Tumor vs normal probability\n(allele)')
print((p1+p2 +p3)+plot_layout(ncol=2))

# Tumor phylogeny
nb$plot_sc_tree(
  label_size = 3, 
  branch_width = 0.5, 
  tip_length = 0.5,
  pal_clone = mypal,
  tip = TRUE)

# mutational history
nb$plot_mut_history(pal=mypal)
dev.off()


system(paste0("aws s3 sync data/",pat,"/ s3://melanoma-ribas/numbat/",pat,"/ --exclude '*_cnv.rds' --quiet"))
print(paste('Finished:', pat))
